mongo-shell-intermediate-read-review (#7)
mattdavis0351/mongodb-labs: exercises/mongo-shell/intermediate-read.md

## Comparison Operators

> These operators allow us to match a field's value relative to another value. 

Reviewer: I think we should explain what **Comparison Operators** are.  Remember that these students are very new to this stuff, having only completed undergrad, so it's helpful to see this type of thing as a review for them.

Additionally, each language seems to have it's own operators, so a quick overview of what MongoDB has isn't a bad thing.

Lastly, MongoDB has **two** set's of **Comparison Operators**, one set for normal queries, and another for aggregate queries.  We need to make a distinction here somehow.

```suggestion
Comparison operators are used in conditions that compare one expression with another, typically containing a boolean return value.  

The MongoDB query language is unique because it uses **comparison Operators** in two different ways:  
- Comparison Query Operators
- Comparison Aggregation Operators

Let's take a quick :eyes: at what is available under each context before moving forward.

**Comparison Query Operators**
|Name|Description|
|---|---|
|$eq|Matches values that are equal to a specified value.|
|$gt | Matches values that are greater than a specified value.|
|$gte | Matches values that are greater than or equal to a specified value.|
|$in | Matches any of the values specified in an array.|
|$lt | Matches values that are less than a specified value.|
|$lte | Matches values that are less than or equal to a specified value.|
|$ne | Matches all values that are not equal to a specified value.|
|$nin | Matches none of the values specified in an array.|

**Comparison Aggregation Operators**
|Name|Description|
|---|---|
|$eq|Returns **true** if the values are equivalent.|
|$gt | Returns **true** if the first value is greater than the second.|
|$gte |Returns **true** if the first value is greater than or equal to the second.|
|$lt | Returns **true** if the first value is less than the second.|
|$lte | Returns **true** if the first value is less than or equal to the second.|
|$ne | Returns **true** if the value are *not* equivalent.|
|$cmp | Returns: 0 if the two values are equivalent, 1 if the first value is greater than the second, and -1 if the first value is less than the second.|

`$cmp` is the only **Comparison Aggregation Operator** that does not return a boolean!
```

Author: This document pertains to read operations at an intermediate level. I figured I'd put the aggregation pipeline in the advanced read document. 

```suggestion

Comparison operators are used in conditions that compare one expression with another, typically containing a boolean return value.  

The MongoDB query language is unique because it uses **comparison Operators** in two different ways:  
- Comparison Query Operators
- Comparison Aggregation Operators

Let's take a quick :eyes: at what is available for **Query Selector** operations.

**Comparison Query Operators**
|Name|Description|
|---|---|
|$eq|Matches values that are equal to a specified value.|
|$gt | Matches values that are greater than a specified value.|
|$gte | Matches values that are greater than or equal to a specified value.|
|$in | Matches any of the values specified in an array.|
|$lt | Matches values that are less than a specified value.|
|$lte | Matches values that are less than or equal to a specified value.|
|$ne | Matches all values that are not equal to a specified value.|
|$nin | Matches none of the values specified in an array.|
```

### Working with a single field

**Exercise 1** :computer: 

In the following query, we look at movies which had a runtime greater than 180 minutes (3 hours.)

**query:**

```javascript=
db.movieDetails.find({"runtime": {"$gt":180}}, 
    {"title": 1, "_id": 0})
```

**Exercise 2** :computer: 

What about movies which ran for greater than or equal to 2 hours but less than 3 hours?

**query:**
```javascript=
db.movieDetails.find({"runtime": {"$gte":120, "$lt": 180}}, 
    {"title": 1, "_id": 0})
```
---

### Working with multiple fields

**Exercise 3** :computer: 

> List the movies which not only had a runtime greater than 3 hours but were also highly rated.

Reviewer: What constitutes a high rating?
I also think it might be wise to get them used to words like `longer` and `higher` being associated with `greater`

```suggestion
List the movies which had a runtime longer than 3 hours and had a rating higher than 8
```

**query:**
```javascript=
db.movieDetails.find({"runtime": {"$gt": 180},
    "imdb.rating": {"$gt":8}}, 
    {"title": 1, "_id": 0})
```
---

> ### The "$ne" operator

Reviewer: ```suggestion
### The "$ne" query operator
```

`ne` stands for **not equal**. Therefore the use of this operator filters out records where the matching condition is not true.

**Exercise 3** :computer: 

Filter out records where the `rating` field does not contain the value of 'UNRATED'.   

**query:**
```javascript
db.movieDetails.find({"rated": {"$ne": "UNRATED"}}, 
    {"_id":0, "title":1, "rated":1})
```
**result:** (Only a part of the output has been shown.)
```javascript
{ "title" : "Zathura: A Space Adventure", "rated" : "PG" }
{ "title" : "Space Cowboys", "rated" : "PG-13" }
{ "title" : "Lost in Space", "rated" : "PG-13" }
{ "title" : "Muppets from Space", "rated" : "G" }
{ "title" : "Turks in Space", "rated" : null }
```

Look at the last document in the output above. The `rated` field has a `null` value. This is because `$ne` **only** excludes the documents where the `rated` field has a value of `UNRATED`. All other values, (be it `APPROVED`, `R`, `PG-13`, `PG`, `null` etc) find their way into the result. In the same vein, `$ne` also outputs documents which do **not** have a `rated` field at all. 

In a future exercise, we'll see how to find documents which do not have a given field. 

---

> ### The "$in" operator

Reviewer: ```suggestion
### The "$in" query operator
```

The `$in` operator allows us to specify 1 or more values, any 1 of which if matched, causes a resulting document to be returned. Note that the value of `$in` should always be an array. 

**Exercise 4** :computer: 

Find movies which were either rated `R`, `PG` or `PG-13`.

**query:**
```javascript
db.movieDetails.find({"rated": {"$in": ["R", "PG", "PG-13"]}}, 
    {"_id":0, "title":1, "rated":1})
```

**Exercise 5** :computer: 

> Use the `$nin` operator (**not in** operator) to match **none** of the values specified in the array. 

Reviewer: ```suggestion
Use the `$nin` query operator (**not in** operator) to match **none** of the values specified in the array. 
```

---

> ---

Reviewer: why two?
```suggestion
```

> ## Element Operators

Reviewer: I'm not sure how we got from **Comparison Query Operators** to here.  Do you want to add in **Comparison Aggregation Operators** prior to getting here, or is the goal of this document to be about **Query Selectors**?

If that's the case, then let's make that more clear.

Author: The goal is Query Selectors. We can add a header topic at the beginning of the document to make that more clear.

Reviewer: okay 👍

    *Hint: use `{"$exists": false}`*

* Now add up the results obtained in exercises 1 and 2. Is it the same as the total collection size?

* Take 5 minutes :alarm_clock: to explore the document structure with and without these fields using the `find()` method. 

* Is the output of the following query the same as the one with `{"$exists": false}`? 

    **query:**

    ```javascript
    db.movieDetails.count({"tomato.consensus": null})
    ```
    
:arrow_right: Note that the query above returns not only documents where the `tomato.consensus` field is `null` but also those documents where the `tomato.consensus` field is absent. Check it out for yourself!

**Exercise 3** :computer: 

Explore the various datatypes present per field using the `$type` operator. Refer to the [documentation](https://docs.mongodb.com/manual/reference/operator/query/type/#op._S_type) for additional reference. 

To help you get started, here's an example:

**query:**
```javascript
db.movieDetails.find({"plot": {"$type": "string"}}).pretty()
```

**query:**
```javascript
db.movieDetails.find({"plot": {"$type": "null"}}).pretty()
```

> ---

Reviewer: What type of lab ⏰ activities can you think of to add to this section? it would be important to have them break into groups and explore these steps to get used to this type of stuff.

> ---

Reviewer: This should probably be removed, maybe even the line above it as well.  Does it make sense to have this weird line at the bottom of the file? 🤷‍♂ 

```suggestion
```

* is `boxOffice.country` equal to "North America" :white_check_mark:
* is `boxOffice.revenue` greater than $400m :white_check_mark:

Both checks were satisfied at **different array positions**!!!

Here's where `$elemMatch` comes in. It ensures that array elements present **in the same position** are evalauted. 

Try the following 2 queries: 
```javascript
db.movieDetails.find({"boxOffice": 
    {"$elemMatch": 
        {"country": "North America", "revenue": 
            {"$gt": 200}}}}).pretty()
            
db.movieDetails.find({"boxOffice": 
    {"$elemMatch": 
        {"country": "North America", "revenue": 
            {"$gt": 400}}}}).pretty()
```

See the difference?!

---

### The "$size" query operator

As the name suggests, this operator deals with the length of an array. 

**Exercise 3** :computer: 

* Which movies were filmed in just one country?

    **query:**

    ```javascript
    db.movieDetails.find({"countries": {"$size": 1}}, 
        {"_id": 0, "title": 1, "countries": 1})
    ```
    
* Similar to the previous query, find the maximum number of countries that a movie(s) was shot in. It'll take a little trial and error but code away! 

> ---

Reviewer: Adding a placeholder for the Activities at the end.  This will need to be edited for the intermediate use cases 

```suggestion
---

## Group Activities
During this time the class should split into :five: groups to complete each section.

**Lab 1: :alarm_clock: 10 minutes** 

Share with the rest of the class the kind of movies you like by writing 3 interesting queries against this database, at the same time limiting the resulting documents to fields that you find relevant. 

For example, movies that were part of both the 'Action' and 'Adventure' genres! :boom: 

**Lab 2: :alarm_clock: 15 minutes**
Using `some other dbs we have` break into small groups and explore the documents in this database.  Create a small 2-3 slide long presentation about the data.  

Some questions to consider while exploring:
- How is this data structured?
- Is there anything interesing to me about this data?
- Is the data malformed, incorrect or incomplete?
- What are the challenges you faced while exploring this data?

**Lab 3: :alarm_clock: 15 minutes**
Present your results from **Lab 2** and answer any questions your classmates may have.
```

Author: ```suggestion
---
## Group Activities
During this time the class should split into :five: groups to complete each section.

**Activity 1: :alarm_clock: 15 minutes** 

Let's figure out some weekend plans! Get together with the rest of your group and look at movies from a genre(s) you all like. Pick a couple of highly rated movies that you haven't watched yet and let the rest of the class know that these will be on your list! Also, are there some medium or low rated movies from that genre(s) that you've already watched? Do you agree with the rating? Why?

**Activity 2: :alarm_clock: 15 minutes**

Explore the different datasets available on your workstation.  Be creative and try to find interesting things.  
Some questions to consider while exploring:
- How is this data structured?
- Is there anything interesting to me about this data?
- Is the data malformed, incorrect or incomplete?
- What are the challenges you faced while exploring this data?

**Presentation: :alarm_clock: 25 minutes**
Present your results from **Group Activities** and answer any questions your classmates may have.
```

Reviewer: I'll go ahead and commit these changes.  I'm okay with the added Activity, I like that you kept the exploration of the other data in here.  

What I said in the other PR still stands... You're leaning on the `movies` db too much. 

Imagine if you were in class and you hated movies, but we only did anything with movie data... you'd get bored, especially if you knew there was other data sitting right there and you loved shipwrecks and the titanic 🤷‍♂ 

Variety is key for this class to be interesting

---

## Logical Operators

### The "$or" operator
This operator **OR**s the filters stated in a query and returns documents where atleast 1 of the filters is true. 

**Exercise 1** :computer: 

Write a query to list movies whose `imdb.rating` is greater than 8.5 or `metacritic` rating is greater than 85.

> :arrow_right: Note that the `$or` operator takes an **array** of values. 

Reviewer: I think this will apply to all of these operators in this document, but I'm just now thinking of it.  Showing syntax of the operator prior to showing a demo query would make for quick notes for the students to view during any assessments.

On top of that, we fucking care about quality around here, so get your 💩 together!

```suggestion
:arrow_right: Note that the `$or` operator takes an **array** of values.

Syntax Example:
`{"$or":[{field:condition}, {field:condition}]}`
```

**query:**
```javascript
db.movieDetails.find({"$or": [
    {"imdb.rating": {"$gt": 8.5}}, 
    {"metacritic": {"$gt": 85}}]}, 
        {"_id": 0, "title": 1, "imdb.rating": 1, "metacritic": 1})
```
---

### The "$and" operator

> As you already know, MongoDB implicitly **AND**s the filters when separated by `,` in the query. 

Reviewer: add a link here, refer to the location in the document that explained this earlier, just for quick navigation.  Cross-links are usually always a good idea.

```suggestion
As you [already know](../blob/master/exercises/mongo-shell/basic-crud/basic-read.md#Filtering-on-multiple-fields), MongoDB implicitly **AND**s the filters when separated by `,` in the query. 
```

This makes use of a relative link with github, similar to you navigating your filesystem which is where the `../blob/master/` comes from.  Also **any** heading (h1, h2, h3 and so on) can be used as a link target.  

In this case, your target heading has spaces in it.  When that is the case you need to include `-` in place of the space inside of the link URL.  You can see how I did that above.

**Exercise 2** :computer: 

> Run the following query:

Reviewer: explain some objective or 'why' behind running the following query.  I know you go into detail when talking about the output below, but 'run the following query' is asking the student to blindly follow you.  

```suggestion
To cover this next point we need to generate some output first.  Run the query below to get the output we need. 

:bulb: Can you explain the different operators and parameters in this query without seeing the output?
```

```javascript
db.movieDetails.find({
    "tomato.meter": {"$lt": 50}, 
    "tomato.meter": {"$ne": null}}, 
    {"_id": 0, "title": 1, "tomato.meter": 1})
```
**result:** (Only a part of the result has been shown.)
```javascript
{ "title" : "Once Upon a Time in the West", "tomato" : { "meter" : 98 } }
{ "title" : "A Million Ways to Die in the West", "tomato" : { "meter" : 33 } }
{ "title" : "Wild Wild West", "tomato" : { "meter" : 17 } }
{ "title" : "Slow West", "tomato" : { "meter" : 92 } }
{ "title" : "Journey to the West", "tomato" : { "meter" : 93 } }
```

> Ideally, the 2 filter conditions would be ANDed. But :eyes: closely at the results and you'll see that the first filter condition `tomato.meter": {"$lt": 50}` was ignored. The result contains documents which have `tomato.meter` ratings greater than 50 as well. Why?

Reviewer: ```suggestion
Ideally, the 2 filter conditions would be **AND**ed. But :eyes: closely at the results and you'll see that the first filter condition `tomato.meter": {"$lt": 50}` was ignored. The result contains documents which have `tomato.meter` ratings greater than 50 as well. Why?
```

db.movieDetails.find({"plot": {"$type": "null"}}).pretty()
```
---

## Array Operators

> In the following exercises, we'll look at operators for array fields. 

Reviewer: It might be useful to make `array fields` at the end of this sentence a link to the Mongo docs for **Array Operators**.  This holds true for the other sections as well.  

Maybe once everything is down and can :ship: we will go back through and do this kind of thing.

Author: Refer to issue #10 !

Copy and run the following code in your mongo shell: (Revenue is in million dollars.)
```javascript
martian = db.movieDetails.findOne({"title": "The Martian"})
delete martian._id
martian.boxOffice = [
    {"country": "North America", "revenue": 228.4},
    {"country": "Others", "revenue": 401.7}
]
db.movieDetails.insertOne(martian)
```

What did we just do?!
Simply put:
* we created a variable called `martian`
* added a reference to the document titled `"The Martian"` within that variable
* went on to add another field called `boxOffice` to that variable which turns out to be an **array of objects**
* inserted that variable in our collection as a document of its own

Now when you run:
```javascript
db.movieDetails.find({"title": "The Martian"}).pretty()
```
you'll see that 2 documents are returned. One's the original document and the other is the one we just created with an additional `boxOffice` field. 

The point of this was to add an **array of objects** field to our collection. Now onto the exercise where we'll see how to leverage `$elemMatch` to query this field. 

**Exercise 2** :computer: 

> Take 5 minutes :alarm_clock: to find the movies where the `boxOffice` revenue in "North America" was greater than 

Reviewer: is 5 minutes enough time?  maybe 7-10?  thoughts?

Author: There is only one movie with a `boxOffice` field!!! :stuck_out_tongue_closed_eyes: And that's the one I asked them to create in the code snippet above. So I don't think they'll need more than 5 minutes.

There is no movie in our collection which has a revenue greater than $400m in North America. 

This happened because the query ran the following checks: 
* is `boxOffice.country` equal to "North America" :white_check_mark:
* is `boxOffice.revenue` greater than $400m :white_check_mark:

Both checks were satisfied at **different array positions**!!!

Here's where `$elemMatch` comes in. It ensures that array elements present **in the same position** are evalauted. 

Try the following 2 queries: 
```javascript
db.movieDetails.find({"boxOffice": 
    {"$elemMatch": 
        {"country": "North America", "revenue": 
            {"$gt": 200}}}}).pretty()
            
db.movieDetails.find({"boxOffice": 
    {"$elemMatch": 
        {"country": "North America", "revenue": 
            {"$gt": 400}}}}).pretty()
```

See the difference?!

---

### The "$size" query operator

As the name suggests, this operator deals with the length of an array. 

**Exercise 3** :computer: 

* Which movies were filmed in just one country?

    **query:**

    ```javascript
    db.movieDetails.find({"countries": {"$size": 1}}, 
        {"_id": 0, "title": 1, "countries": 1})
    ```
    

> * Similar to the previous query, find the maximum number of countries that a movie(s) was shot in. It'll take a little trial and error but code away! 

Reviewer: We need a timestamp on this activity!!!  C'mon man! Quality > Quantity :wink:

Author: Done

```suggestion
* Similar to the previous query, take **5 minutes** ⏰ to find the maximum number of countries that a movie(s) was shot in. It'll take a little trial and error but code away! 
```
